### PR TITLE
Modern JS for SSR Bundles

### PR DESCRIPTION
**Pull request by developit**
Next.js already produces fully independent bundles for the server. This is required by `DefinePlugin` and various other things, and generally avoids bundling server code into the client-side application.

This PR exposes Next's internal Boolean `isServer` value, which indicates that a given compilation (or portion thereof) is part of the generation of bundles that will run on the server in a Node.js runtime.

The most notable change here is that a new mechanism has been introduced for passing values like `isServer` to Babel presets & plugins. This is done to ensure custom `.babelrc` configurations don't turn off the desired behavior. It remains possible to compile for a specific browserslist/node/esmodules target, however the default is now SSR-aware.

The benefit of doing this is that the output better leverages modern JavaScript when the current version of Node supports it, leading to faster runtime performance and smaller output size. Both valuable qualities, particularly in serverless deployment scenarios.

Comparing the SSR bundles output of the ["basic" integration test](https://github.com/zeit/next.js/tree/canary/test/integration/basic) shows an expected improvement in bundle sizes when building on Node 10:

| File | Before | After |
| ---: | :---: | :---: |
| `0.*****.js` | 299b | 293b
| `36.*****.js` | 985b | 492b
| `37.*****.js` | 714b | 296b
| `38.*****.js` | 301b | 415b
| `39.*****.js` | 413b | 444b
| `40.*****.js` | 462b | 386b
| `hello-world.*****.js` | 303b | 298b
| `hello1.*****.js` | 300b | 294b
| `hello2.*****.js` | 300b | 294b

**Total:** 4077b to 3212b, a 21% reduction in size.

While these bundles are all extremely small, it's important to note that the scaling characteristics are what matter here, not absolute byte size. This percentage improvement is expected to hold relatively true as applications scale. A 20% savings is also directly in line with the runtime performance improvements @mathiasbynens demonstrated previously, where modern JS output resulted in a 17% increase in SSR throughput.

<details>
<summary><strong>🎉 Updated Benchmarks showing further size improvements</strong></summary>

| File | Before | After |
| ---: | :---: | :---: |
| `static/*****/pages/_document.js` | 8627b | 6098b
| `static/*****/pages/_app.js` | 6813b | 4637b
| `static/*****/pages/dynamic/bundle.js` | 4174b | 2669b
| `static/*****/pages/_error.js` | 3772b | 2147b
| `static/*****/pages/hmr/error-in-gip.js` | 2981b | 1401b
| `36.*****.js` |  985b |  492b
| `37.*****.js` |  714b |  444b
| `40.*****.js` |  462b |  415b
| `39.*****.js` |  413b |  386b
| `hello-world.*****.js` |  303b |  298b
| `38.*****.js` |  301b |  296b
| `hello1.*****.js` |  300b |  294b
| `hello2.*****.js` |  300b |  294b
| `0.*****.js` |  299b |  293b
| `ssr-module-cache.js` |  101b |  101b

</details>

### CODE REVIEW

File: packages/next/build/webpack/loaders/next-babel-loader.js [hunk 1]
```diff
@@ -34,6 +34,8 @@ module.exports = babelLoader.custom(babel => {
             cacheDirectory: join(opts.distDir, 'cache', 'next-babel-loader'),
             cacheIdentifier:
                 cacheKey +
+                // @TODO: this point of variability doesn't seem to be accounted for?
+                // (opts.isServer ? '-server' : '') +
```
Comment (developit) [inline packages/next/build/webpack/loaders/next-babel-loader.js]: @timneutkens @Timer I haven't found a way to check that this is an issue/non-issue yet. Do you know off-hand if the client and server webpack passes share a Babel cache key?

Comment (Timer) [reply packages/next/build/webpack/loaders/next-babel-loader.js]: Off hand, they do share a cache key. We'll need to add the server delimiter here.

File: packages/next/build/babel/preset.ts [hunk 2]
```diff
@@ -59,13 +59,25 @@ module.exports = (
...
     ...options['preset-env'],
   }
+
+  // When transpiling for the server, target the current Node version if not explicitly specified:
+  if (
+    isServer &&
+    (!presetEnvConfig.targets || !('node' in presetEnvConfig.targets))
+  ) {
+    presetEnvConfig.targets = {
+      node: true,
```
Comment (Timer) [inline packages/next/build/babel/preset.ts]:
```suggestion
      // Targets the current process' version of Node. This requires apps be
      // built and deployed on the same version of Node.
      node: 'current',
```

Can we use `current` over `true` for clarity?

**Comment by Timer** [comment]
Also, we should probably test this somehow... Not sure of a good test.

File: packages/next/build/babel/preset.ts (hunk 2 above)
Comment (developit) [reply packages/next/build/babel/preset.ts]: yup, good call.

**Comment by developit** [comment]
Ready for review!

**Review by timneutkens — CHANGES_REQUESTED** [review]
Looks good to me 💯, let's have @Timer do a final review!

**Review by timneutkens — APPROVED** [review]
Pressed the wrong button 😅 Approved.

**Review by Timer — APPROVED** [review]
This looks great to me -- I'm looking forward to the performance boosts.

**Comment by mathiasbynens** [comment]
Super happy to see this land!

With this change, the Node.js version used to build becomes more important. As suggested in [our investigation doc](https://docs.google.com/document/d/1oGV7lrL-DF1pgGeP5Lux2b0FrEAapmDh1-Nec8AWzMU/edit#bookmark=id.9sg6yrf8gako), I'd love to see this nuance added to the docs in some way. In particular, developers should feel encouraged to add and use an `.nvmrc` file to their project, and strive to bump the version regularly. Does anyone want to kick off a documentation PR that recommends this best practice?

**Comment by iamstarkov** [comment]
@mathiasbynens can investigation document be made public?

**Comment by mathiasbynens** [comment]
@iamstarkov Apologies, it was supposed to be public before. It is public now!

**Comment by iamstarkov** [comment]
@mathiasbynens thanks, but it is still not public neither for me, nor for two my colleagues.

**Comment by mathiasbynens** [comment]
@iamstarkov The URL changed.

**Comment by iamstarkov** [comment]
@mathiasbynens it works now. thank you

**Pull request merged**